pallets/parachain-staking/src/weights.rs

```diff
@@ -58,260 +57,256 @@ pub trait WeightInfo {
-		(59_735_000 as Weight)
+		(58_748_000 as Weight)
 			.saturating_add(RocksDbWeight::get().reads(8 as Weight))
 			.saturating_add(RocksDbWeight::get().writes(6 as Weight))
 	}
 	fn candidate_bond_less() -> Weight {
-		(59_421_000 as Weight)
+		(58_578_000 as Weight)
 			.saturating_add(RocksDbWeight::get().reads(8 as Weight))
 			.saturating_add(RocksDbWeight::get().writes(6 as Weight))
 	}
-	fn nominate(x: u32, y: u32) -> Weight {
-		(71_656_000 as Weight)
-			// Standard Error: 1_000
-			.saturating_add((1_049_000 as Weight).saturating_mul(x as Weight))
-			// Standard Error: 5_000
-			.saturating_add((947_000 as Weight).saturating_mul(y as Weight))
+	fn nominate(x: u32, y: u32, ) -> Weight {
+		(85_728_000 as Weight)
+			// Standard Error: 3_000
+			.saturating_add((587_000 as Weight).saturating_mul(x as Weight))
+			// Standard Error: 3_000
+			.saturating_add((637_000 as Weight).saturating_mul(y as Weight))
 			.saturating_add(RocksDbWeight::get().reads(9 as Weight))
 			.saturating_add(RocksDbWeight::get().writes(7 as Weight))
 	}
-	fn leave_nominators(x: u32) -> Weight {
-		(36_354_000 as Weight)
+	fn leave_nominators(x: u32, ) -> Weight {
+		(39_075_000 as Weight)
 			// Standard Error: 2_000
-			.saturating_add((694_000 as Weight).saturating_mul(x as Weight))
+			.saturating_add((374_000 as Weight).saturating_mul(x as Weight))
 			.saturating_add(RocksDbWeight::get().reads(7 as Weight))
 			.saturating_add(RocksDbWeight::get().writes(4 as Weight))
 	}
 	fn revoke_nomination() -> Weight {
-		(37_580_000 as Weight)
+		(36_803_000 as Weight)
 			.saturating_add(RocksDbWeight::get().reads(7 as Weight))
 			.saturating_add(RocksDbWeight::get().writes(4 as Weight))
 	}
 	fn nominator_bond_more() -> Weight {
-		(70_867_000 as Weight)
+		(68_870_000 as Weight)
 			.saturating_add(RocksDbWeight::get().reads(9 as Weight))
 			.saturating_add(RocksDbWeight::get().writes(7 as Weight))
 	}
 	fn nominator_bond_less() -> Weight {
-		(70_857_000 as Weight)
+		(68_980_000 as Weight)
 			.saturating_add(RocksDbWeight::get().reads(9 as Weight))
 			.saturating_add(RocksDbWeight::get().writes(7 as Weight))
 	}
-	// If this takes up too much block space, run again on code
-	// - #743 benchmarks post reward payout optimization were 3x lower per collator,
-	// 15x lower per nominator
-	fn active_on_initialize(x: u32, y: u32) -> Weight {
+	fn active_on_initialize(x: u32, y: u32, ) -> Weight {
 		(0 as Weight)
-			// Standard Error: 299_000
-			.saturating_add((208_550_000 as Weight).saturating_mul(x as Weight))
-			// Standard Error: 27_000
-			.saturating_add((15_580_000 as Weight).saturating_mul(y as Weight))
-			.saturating_add(RocksDbWeight::get().reads(26 as Weight))
+			// Standard Error: 725_000
+			.saturating_add((129_379_000 as Weight).saturating_mul(x as Weight))
+			// Standard Error: 6_000
+			.saturating_add((840_000 as Weight).saturating_mul(y as Weight))
+			.saturating_add(RocksDbWeight::get().reads(101 as Weight))
```

Comment:
@4meta5 Is this the change we saw earlier that we decided to ignore by not updating the banchmarks? Should we do the same here?

Reply:
Yes, it was active_on_initialize. I don't see why we shouldn't update it, but I also never understood why it wasn't updated in the PR that contained the optimization.

Reply:
This change actually increases the weight a lot -- there are a lot more reads/writes than previous. So maybe this doesn't have to do with the change I was initially talking about. :thinking: 

Anyway, I'm a fan of having `on_initialize` tally up its weight on the fly like as we've discussed. In that case, we don't even really need a benchmark for it...

Reply:
It increases number of reads/writes but decreases the complexity parameter coefficients which are the number of collator candidates and nominators. These dominate the cost of the benchmark more than the reads and writes number. 

Especially because the PoV size is the main constraint and for the PoV it is entirely in memory so there are no db reads/writes.

How would you tally up the weight? I've never seen that approach. We shouldn't ignore nor conservatively estimate cost of the computation which is what we usually do for migrations. The purpose of these weight functions is to estimate the average cost equation which should be dominated by the number of collator candidates and number of delegators.

Reply:
Oh, you're right. That would be a decrease in real-world cases.

Here's an idea for an approach to weights in `on_initialize()`:

1. Benchmark all of it that is *directly* related to number of collators/nominators, use this as a baseline for weight
2. For the variable parts that take advantage of your recent optimizations, tally the operations up manually. This might be as simple as having `pay_stakers()` return its own weight and add it to `1` above. This could be done in code but could probably also rely on benchmarking.